src/datalogging.cpp

```diff
@@ -1,108 +1,170 @@
-        log.close();
+    if (!sd.begin(cfg)) {
+        Serial.print("\tFS mount failed sdErr=0x");
+        Serial.print(sd.sdErrorCode(), HEX);
+        Serial.print(" sdData=0x");
+        Serial.println(sd.sdErrorData(), HEX);
+        return;
     }
+    Serial.println("\tFS mounted");
+
+    // Bootcount
+    bootcount = readUintFromFile(BOOTCOUNT_FILE, 0) + 1;
+    writeUintToFile(BOOTCOUNT_FILE, bootcount);
 
-    strncpy((char*)CSV_FILE, indexedLogFile, FILE_NAME_SIZE);
+    // Log file name
+    char indexedLogFile[FILE_NAME_SIZE] = {0};
+    snprintf(indexedLogFile, FILE_NAME_SIZE, "log_%lu.csv",
+             (unsigned long)bootcount);
+    appendCsvHeaderIfNew(indexedLogFile);
+    strncpy(CSV_FILE, indexedLogFile, FILE_NAME_SIZE);
 
-    Serial.println("Datalogging initialized");
-    Serial.print("Boot count: ");
+    Serial.print("\tBoot count: ");
     Serial.println(bootcount);
-    Serial.print("Log file: ");
+    Serial.print("\tLog file: ");
     Serial.println(CSV_FILE);
 }
 
-void dataloggingExecute() {
-    GnssData gnssData = {};
-    SensingData sensingData = {};
-    
-    xQueueReceive(gnssQueue, &gnssData, 0);
-    xQueueReceive(sensingQueue, &sensingData, 0);
-    // Write to CSV
-    File log = fatfs.open(CSV_FILE, FILE_WRITE);
-    if (log) {
-        log.print(gnssData.time);
-        log.print(",");
-        log.print(gnssData.latitude, 6);
-        log.print(",");
-        log.print(gnssData.longitude, 6);
-        log.print(",");
-        log.print(gnssData.altitude, 2);
-        log.print(",");
-        log.print(sensingData.humidity, 2);
-        log.print(",");
-        log.print(sensingData.temperature, 2);
-        log.print(",");
-        log.print(sensingData.pressure, 2);
-        log.print(",");
-        log.print(sensingData.baroAltitude, 2);
-        log.println();
-        log.close();
+void dataloggingExecute(const GnssData& gnssData, const SensingData& sensingData) {
+    if (!CSV_FILE[0]) {
+        Serial.println("CSV_FILE not set, skipping datalogging");
+        return;
     }
+
+    FsFile log = sd.open(CSV_FILE, O_WRITE | O_CREAT | O_APPEND);
+    if (!log) return;
+
+    log.print(millis());
+    log.print(',');
+    log.print(gnssData.time);
+    log.print(',');
+    log.print(gnssData.latitude, 6);
+    log.print(',');
+    log.print(gnssData.longitude, 6);
+    log.print(',');
+    log.print(gnssData.altitude, 2);
+    log.print(',');
+    log.print(sensingData.baroAltitude, 2);
+    log.print(',');
+    log.print(sensingData.pressure, 2);
+    log.print(',');
+    log.print(sensingData.temperature, 2);
+    log.print(',');
+    log.print(sensingData.humidity, 2);
+    log.println();
+    log.close();
 }
 
 void dataloggingTask(void* pvParameters) {
     Serial.println("Datalogging task started");
+
+    // TODO: If this is ever used. Initialize them with xQueueCreate
+    QueueHandle_t gnssQueue = nullptr;
```

Comment:
for these two, who else is going to initialize them. They're local to the function so it cant be anyone else, and yet you aren't initializing them here. 
And if you initialized them here, no one could send down them. 
Kudos for having the null check so it doesnt crash but i dont get what the declarations are doin ghere

Reply:
Would have to be initialized globally. I just moved them here instead of global for silliness since they aren't being used atm, but might make more sense to just make global, so its once less thing for people to do if they want to RTOS in the future